newrelic/api/opentelemetry.py

```diff
@@ -0,0 +1,476 @@
+        if (
+            not self.otel_parent
+            or (self.otel_parent and self.otel_parent.span_id == int(current_nr_trace.guid, 16))
+            or (self.otel_parent and isinstance(current_nr_trace, Sentinel))
+        ):
+            # Expected to come here if one of three scenarios have occured:
+            # 1. `start_as_current_span` was used.
+            # 2. `start_span` was used and the current span was explicitly set
+            #    to the newly created one.
+            # 3. Only a Sentinel Trace exists so far while still having a
+            #    remote parent. From OTel's end, this will be represented
+            #    as a `NonRecordingSpan` (and be seen as `None` at this
+            #    point). This covers cases where span is remote.
+            self.nr_parent = current_nr_trace
+        else:
+            # Not sure if there is a usecase where we could get in here
+            # but for debug purposes, we will raise an error
+            _logger.warning(
+                "OpenTelemetry span (%s) and NR trace (%s) do not match nor correspond to a remote span.",
+                self.otel_parent,
+                current_nr_trace,
+            )
+            raise ValueError("Unexpected span parent scenario encountered")
+
+        if nr_trace_type == FunctionTrace:
+            trace_kwargs = {"name": self.name, "params": self.attributes, "parent": self.nr_parent}
+            self.nr_trace = nr_trace_type(**trace_kwargs)
+        elif nr_trace_type == DatastoreTrace:
+            trace_kwargs = {
+                "product": self.instrumenting_module,
+                "target": None,
+                "operation": self.name,
+                "parent": self.nr_parent,
+            }
+            self.nr_trace = nr_trace_type(**trace_kwargs)
+        elif nr_trace_type == ExternalTrace:
+            trace_kwargs = {
+                "library": self.name or self.instrumenting_module,
+                "url": self.attributes.get("http.url"),
+                "method": self.attributes.get("http.method"),
+                "parent": self.nr_parent,
+            }
+            self.nr_trace = nr_trace_type(**trace_kwargs)
+        elif nr_trace_type == MessageTrace:
+            trace_kwargs = {
+                "library": self.instrumenting_module,
+                "operation": "Produce",
+                "destination_type": "Topic",
+                "destination_name": self.name,
+                "params": self.attributes,
+                "parent": self.nr_parent,
+                "terminal": False,
+            }
+            self.nr_trace = nr_trace_type(**trace_kwargs)
+        else:
+            # TODO: Still need to implement GraphQLOperationTrace and GraphQLResolverTrace
+            trace_kwargs = {"name": self.name, "params": self.attributes, "parent": self.nr_parent}
+            self.nr_trace = nr_trace_type(**trace_kwargs)
+
+        self.nr_trace.__enter__()
+
+    def _is_sampled(self):
```

Comment:
Are these `_is_sampled` and `_is_remote` functions part of the otel api? I don't see them listed and if they aren't we should make these names more Pythonic (aka: `sampled`, `remote`). The prefixing of `is` comes from javascript (maybe other languages do this too but not generally Python).